src/typechecker.erl

```diff
@@ -1076,6 +1091,15 @@ merge_union_types(Types, _Env) ->
 is_top(?top()) -> true;
 is_top(_) -> false.
 
+%% Decides whether the type expresses exactly one value.
+-spec is_singleton_type(type()) -> boolean().
+is_singleton_type({atom, _, _Atom}) -> true;
+is_singleton_type({integer, _, _Integer}) -> true;
+is_singleton_type({char, _, _Char}) -> true;
+is_singleton_type(?type(range, [{integer, _, N}, {integer, _, N}])) -> true;
+is_singleton_type(?type(map, [])) -> true;
+is_singleton_type(_) -> false.
```

Comment:
The empty list (nil) is also a singleton type. The empty binary too.

Reply:
You're right! I'll add the case for `nil`. What's the empty binary type? `{type,Ann,binary,[{integer,0,0},{integer,0,0}]}`?

Reply:
Exactly:

```
$ nl -b a z.erl
     1	-module(z).
     2
     3	-type z() :: <<>>.
$ erlc +dabstr z.erl
z.erl:3:2: Warning: type z() is unused
%    3| -type z() :: <<>>.
%     |  ^

$ cat z.abstr
{attribute,{1,1},file,{"z.erl",1}}.
{attribute,{1,2},module,z}.
{attribute,{3,2},
           type,
           {z,{type,{3,14},binary,[{integer,3,0},{integer,3,0}]},[]}}.
{eof,{4,1}}.
```

Reply:
Great, I'll add that as well. :)

Reply:
Done.

Reply:
Wait, there's more:

The empty tuple and any tuple where all elements are singleton types. :) Don't know how far we should go, but I guess there are more of these, e.g. maps of only singletons, etc.

Reply:
Good spot!

We probably could go to great lengths to define this function precisely but I think it's not needed now. At least until it's used somewhere else. IMHO, for the use in maps (to catch when a map contains non-singular required fields and disable exhaustiveness checking in that case), the first three clauses are enough (when considering practical use cases), with the first one being by far the most used.

Maybe we should also add the empty tuple type (when we already have the empty map type), but I wouldn't extend this function to be recursive until we really need it. What do you think?

Reply:
I guess it's fine for now but in the long run it's be good if the typechecker is predictable, i.e. either does exhaustiveness checking whenever it's possible or it stops at some predictable depth or so.

Reply:
Ok, I added the case for the empty tuple and a TODO note to add more cases.